### PR TITLE
BuildRustCrate: Fixing compilation issue on Darwin

### PR DESCRIPTION
**Pull request by P-E-Meunier**
###### Motivation for this change
BuildRustCrate was broken on Darwin because of `dontStrip` (see #27370). This PR also updates BuildRustCrate for a number of other things that were checked in the process of debugging.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/build-support/rust/build-rust-crate/default.nix [hunk 1]
```diff
@@ -22,7 +22,10 @@ let
           else
             extern;
         in (if lib.lists.any (x: x == "lib") dep.crateType then
-           " --extern ${name}=${dep.lib}/lib/lib${extern}-${dep.metadata}.rlib"
+           if linked then
+              " --extern ${name}=${dep.lib}/lib/lib${extern}-${dep.metadata}.rlib"
+	   else
```
Comment (matthewbauer) [inline pkgs/build-support/rust/build-rust-crate/default.nix]:
```suggestion
           else
```

File: pkgs/build-support/rust/build-rust-crate/default.nix [hunk 2]
```diff
@@ -22,7 +22,10 @@ let
...
+	      " --extern ${name}=${dep.lib}/lib/lib${extern}-${dep.metadata}.rmeta"
```
Comment (matthewbauer) [inline pkgs/build-support/rust/build-rust-crate/default.nix]:
```suggestion
             " --extern ${name}=${dep.lib}/lib/lib${extern}-${dep.metadata}.rmeta"
```

File: pkgs/build-support/rust/build-rust-crate/default.nix [hunk 3]
```diff
@@ -94,6 +97,7 @@ stdenv.mkDerivation (rec {
     name = "rust_${crate.crateName}-${crate.version}";
     depsBuildBuild = [ rust stdenv.cc ];
     buildInputs = (crate.buildInputs or []) ++ buildInputs_;
+    dontStrip = stdenv.hostPlatform.isDarwin;
```
Comment (matthewbauer) [inline pkgs/build-support/rust/build-rust-crate/default.nix]: could you add a comment linking to the issue in case we ever revisit this

**Comment by Profpatsch** [comment]
(triage) is this still relevant?

**Comment by P-E-Meunier** [comment]
Yes, I just fixed the comments (thanks @matthewbauer, and sorry for the delay).

**Comment by andir** [comment]
Can you give us some kind of description for those various other things that you added? The `.rmeta` addition (which is now the default) has what kind of consequences? I haven't really found a use-case for them thus I am eager to learn about it.

**Comment by P-E-Meunier** [comment]
Actually, I don't really know! The `.rmeta` was added in an attempt to mimic more closely what cargo does now (as opposed to what it was doing when carnix was started). The rmetas are supposed to speed up compilation, although I have no data about this, but might be a good idea to do in the future anyway.

I introduced them before I found out about `dontStrip`. I haven't been using Carnix too much recently, I could maybe try to benchmark it (and possibly split this PR).

**Comment by andir** [comment]
> Actually, I don't really know! The `.rmeta` was added in an attempt to mimic more closely what cargo does now (as opposed to what it was doing when carnix was started). The rmetas are supposed to speed up compilation, although I have no data about this, but might be a good idea to do in the future anyway.
> 
> I introduced them before I found out about `dontStrip`. I haven't been using Carnix too much recently, I could maybe try to benchmark it (and possibly split this PR).

My understanding is that those files are useful for cargo when buildings dependencies in parallel. Since we have them properly isolated and the processes do not know anything about each other we probably shouldn't take that route.

**Comment by P-E-Meunier** [comment]
I believe the rmeta are supposed to be barebones specifications of the library, which might potentially result in smaller closures.

**Comment by andir** [comment]
> I believe the rmeta are supposed to be barebones specifications of the library, which might potentially result in smaller closures.

I am tempted to give them a go. I currently have a rare issue where linking multiple `.rlib` files from different builders results in metadata hashsum mismatches. Maybe that fixes it.

**Pull request closed without merging**